File: src/main/java/me/kaotich00/fwauctionhouse/message/Message.java
package me.kaotich00.fwauctionhouse.message;

import me.kaotich00.fwauctionhouse.FwAuctionHouse;
import me.kaotich00.fwauctionhouse.locale.LocalizationManager;
import me.kaotich00.fwauctionhouse.utils.MessageUtils;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.command.CommandSender;

public enum Message {

    PREFIX("fwauctionhouse_prefix", false),

    HELP_MESSAGE("help_message", false),
    SOLD_ITEM("sold_item", true),
    CANNOT_SELL_AIR("cannot_sell_air", true);

    private final String message;
    private final boolean showPrefix;
    private final LocalizationManager localizationManager;

    Message(String message, boolean showPrefix) {
        this.message = MessageUtils.rewritePlaceholders(message);
        this.showPrefix = showPrefix;
        this.localizationManager = FwAuctionHouse.getLocalizationManager();
    }

    public void send(CommandSender sender, Object... objects) {
        sender.sendMessage(asString(objects));
    }

    public void broadcast(Object... objects) {
        Bukkit.getServer().broadcastMessage(asString(objects));
    }

    public String asString(Object... objects) {
        return format(objects);
    }

    private String format(Object... objects) {
        String string = localizationManager.localize(this.message);
        if(this.showPrefix) {
            string = localizationManager.localize(PREFIX.message) + " " + string;
        }
        for (int i = 0; i < objects.length; i++) {
            Object o = objects[i];
            string = string.replace("{" + i + "}", String.valueOf(o));
        }
        return ChatColor.translateAlternateColorCodes('&', string);
    }

    public static LocalizationManager getLocalizationManager() {
        return LocalizationManager.getInstance(FwAuctionHouse.getPlugin(FwAuctionHouse.class));
    }

}

File: src/main/java/me/kaotich00/fwauctionhouse/message/Message.kt
package me.kaotich00.fwauctionhouse.message

import me.kaotich00.fwauctionhouse.locale.LocalizationManager
import net.kyori.adventure.audience.Audience
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.TextComponent
import net.kyori.adventure.text.format.NamedTextColor
import org.bukkit.command.CommandSender

enum class Message(
    private val showPrefix: Boolean = true
) {
    HELP_MESSAGE(false),
    SOLD_ITEM,
    BOUGHT_ITEM,
    INVENTORY_FULL,
    NOT_ENOUGH_MONEY,
    CANNOT_SELL_AIR,
    PURCHASE_MESSAGE(false),
    VALIDATED_TOKEN,
    VALIDATED_TOKEN_MESSAGE,
    DECLINED,
    ENTER_BUILD_MODE,
    ALREADY_IN_BUILD_MODE,
    NOT_IN_BUILD_MODE,
    EXIT_BUILD_MODE,
    FIRST_POSITION_SET,
    SECOND_POSITION_SET,
    MARKET_AREA_CREATED,
    MARKET_AREA_DELETED,
    NO_SUCH_MARKET_AREA,
    POSSIBLE_OPTIONS_ENTER_EXIT;


    fun send(audience: Audience, vararg params: Any) {
        audience.sendMessage(asComponent(*params))
    }

    fun asComponent(vararg params: Any) = format(params)

    private fun format(params: Array<out Any>): Component {
        val localized = requireLocalizationManager().localize(name, params)
        return if (showPrefix) {
            PREFIX.append(localized)
        } else {
            localized
        }
    }

    private fun requireLocalizationManager() =
        localizationManager ?: error("LocalizationManager not set")

    companion object {

        private val PREFIX = TextComponent.ofChildren(
            Component.text("[", NamedTextColor.DARK_GRAY),
            Component.text("Fw", NamedTextColor.GRAY),
            Component.text("Market", NamedTextColor.GREEN),
            Component.text("] ", NamedTextColor.DARK_GRAY)
        )

        var localizationManager: LocalizationManager? = null
    }
}